README
This tool is used to send messages between the system and users'
accounts.

For each X11 session a small program called s2u is launched under an
X11 dbus session (using existing one if possible). The s2u program
waits then for dbus messages to take the appropriate actions. The
implemented actions are:

 o react to hostname changes to continue to have a working X11 session.
 o run update-menus if requested by the system
 o display msec notifications


ChangeLog
2009-03-23 15:48  Eugeni Dodonov <eugeni at mandriva.com>

	* Makefile, s2u.c: Added support for msec notifications.

2009-02-10 14:02  Eugeni Dodonov <eugeni at mandriva.com>

	* s2u.c: Removed trailing spacing.

2009-02-10 13:58  Eugeni Dodonov <eugeni at mandriva.com>

	* README: Updated README with missing functionality.

2007-02-20 16:17  Frederic Crozat <fcrozat at mandriva.com>

	* Makefile: Add svnroot

2007-02-20 16:16  Frederic Crozat <fcrozat at mandriva.com>

	* Makefile: Fix naming for xinit.d, not a problem since we rely on
	  system bus

2007-02-20 16:15  Frederic Crozat <fcrozat at mandriva.com>

	* Makefile, s2u.sh, s2u.spec: -Release 0.8
	  - fix Makefile to be svn compliant
	  - remove --debug from xinit.d file
	  - rename xinit.d file to be sure it is started after dbus

2006-08-04 12:01  Frederic Crozat <fcrozat at mandriva.com>

	* s2u.spec: Fix buildrequires

2006-08-01 09:24  Frederic Crozat <fcrozat at mandriva.com>

	* ChangeLog: Generated by cvs2cl the 01_Aug

2006-08-01 09:20  Frederic Crozat <fcrozat at mandriva.com>

	* s2u.spec: Rebuild with latest dbus

2006-05-29 14:21  Frederic Crozat <fcrozat at mandriva.com>

	* ChangeLog: Generated by cvs2cl the 29_May

2006-05-29 14:19  Frederic Crozat <fcrozat at mandriva.com>

	* s2u.spec: Prevent crash when parsing cookie fails

2006-05-29 14:19  Frederic Crozat <fcrozat at mandriva.com>

	* s2u.c: Prevent crash when parsing cookie fail

2006-05-29 12:29  Frederic Crozat <fcrozat at mandriva.com>

	* ChangeLog: Generated by cvs2cl the 29_May

2006-05-29 12:29  Frederic Crozat <fcrozat at mandriva.com>

	* Makefile, s2u.c, s2u.spec: - Fix for modular xorg

2006-01-27 21:54  Thierry Vignaud <tvignaud at mandriva.com>

	* ChangeLog: Generated by cvs2cl the 27_Gen

2006-01-27 21:54  Thierry Vignaud <tvignaud at mandriva.com>

	* s2u.spec: bump 0.6-1mdk's date in order to make it acceptable by
	  rpm

2006-01-27 21:52  Thierry Vignaud <tvignaud at mandriva.com>

	* s2u.spec: merge in changes done in cooker and not in CVS

2005-12-16 14:29  Frederic Lepied

	* ChangeLog: Generated by cvs2cl the 16_Dec

2005-12-16 14:28  Frederic Lepied

	* AUTHORS, hostname-post, s2u.conf: Mandriva

2005-12-16 14:27  Frederic Lepied

	* s2u.spec: 0.6-1mdk

2005-12-16 14:26  Frederic Lepied

	* s2u.sh: - removed unneeded code
	  - launch with debug

2005-12-16 14:25  Frederic Lepied

	* Makefile: fixed version/release parsing

2005-12-16 14:24  Frederic Lepied

	* s2u.c: - added --debug option
	  - removed dead code for opt_run_as
	  - Mandriva

2005-03-09 15:30  Frederic Crozat <fcrozat at mandriva.com>

	* ChangeLog: Generated by cvs2cl the 09_Mar

2005-03-09 15:30  Frederic Crozat <fcrozat at mandriva.com>

	* s2u.c, s2u.spec: Remove dbus-x11 dependency

2005-03-09 15:09  Frederic Crozat <fcrozat at mandriva.com>

	* ChangeLog: Generated by cvs2cl the 09_Mar

2005-03-09 15:08  Frederic Crozat <fcrozat at mandriva.com>

	* s2u.c, s2u.spec: - add new signal to start update-menus if
	  requested by system

2005-03-08 16:22  Frederic Crozat <fcrozat at mandriva.com>

	* ChangeLog: Generated by cvs2cl the 08_Mar

2005-03-08 16:22  Frederic Crozat <fcrozat at mandriva.com>

	* Makefile, s2u.c, s2u.spec: Connect to X so s2u exits when X exit

2005-03-07 16:11  Frederic Crozat <fcrozat at mandriva.com>

	* ChangeLog: Generated by cvs2cl the 07_Mar

2005-03-07 16:11  Frederic Crozat <fcrozat at mandriva.com>

	* hostname-post: Fix recipient

2005-03-07 16:04  Frederic Crozat <fcrozat at mandriva.com>

	* ChangeLog: Generated by cvs2cl the 07_Mar

2005-03-07 16:04  Frederic Crozat <fcrozat at mandriva.com>

	* Makefile: Fix perms for config file

2005-03-07 16:01  Frederic Crozat <fcrozat at mandriva.com>

	* ChangeLog: Generated by cvs2cl the 07_Mar

2005-03-07 16:01  Frederic Crozat <fcrozat at mandriva.com>

	* Makefile, hostname-post, s2u.c, s2u.sh, s2u.spec: Release 0.5 :
	  no longer use session bus, use system bus instead (fix Mdk bug
	  #13166)

2005-03-03 18:49  Frederic Crozat <fcrozat at mandriva.com>

	* s2u.conf: Configuration to restrict s2u message to root only on
	  system bus

2004-10-18 15:26  Gwenole Beauchesne

	* ChangeLog: Generated by cvs2cl the 18_Oct

2004-10-18 15:23  Gwenole Beauchesne

	* s2u.spec: 0.4-1mdk

2004-10-18 15:22  Gwenole Beauchesne

	* Makefile: use pkgconfig to get the right includes/libs for
	  dbus-glib-1

2004-10-18 15:22  Gwenole Beauchesne

	* s2u.c: #include <dbus/dbus-glib-lowlevel.h> for
	  dbus_connection_setup_with_g_main()
	  prototype

2004-08-25 05:26  Frederic Lepied

	* s2u.spec: 0.3-1mdk

2004-08-25 05:25  Frederic Lepied

	* s2u.sh: use display in temp filename

2004-08-25 05:24  Frederic Lepied

	* s2u.spec: 0.3-1mdk

2004-08-18 02:56  Frederic Lepied

	* ChangeLog: Generated by cvs2cl the 18_Aug

2004-08-18 02:56  Frederic Lepied

	* s2u.spec: 0.2-1mdk

2004-08-18 02:56  Frederic Lepied

	* hostname-post, s2u.sh: put session file in
	  /tmp/dbus-$USER-xsession$CURR_DISPLAY

2004-08-01 23:44  Frederic Lepied

	* ChangeLog: Generated by cvs2cl the 02_Aug

2004-08-01 23:44  Frederic Lepied

	* s2u.spec: 0.1-1mdk

2004-08-01 23:44  Frederic Lepied

	* Makefile: network-script => network-scripts

2004-08-01 23:34  Frederic Lepied

	* ChangeLog: Generated by cvs2cl the 02_Aug

2004-08-01 23:29  Frederic Lepied

	* ., .cvsignore: first version

2004-08-01 23:28  Frederic Lepied

	* ChangeLog: Generated by cvs2cl the 02_Aug

2004-08-01 23:27  Frederic Lepied

	* s2u.spec: use $RPM_OPT_FLAGS to compile C files

2004-08-01 23:27  Frederic Lepied

	* Makefile: added AUTHORS

2004-08-01 23:23  Frederic Lepied

	* ChangeLog: Generated by cvs2cl the 02_Aug

2004-08-01 23:21  Frederic Lepied

	* s2u.c: added ident strings

2004-08-01 23:20  Frederic Lepied

	* s2u.c: fixed header and strings to remove the drakxservice
	  references

2004-08-01 23:15  Frederic Lepied

	* ChangeLog: Generated by cvs2cl the 02_Aug

2004-08-01 23:08  Frederic Lepied

	* s2u.spec: added a dependency on initscripts


s2u.c
/*===========================================================================
 * Project         : Mandriva Linux
 * Module          : s2u
 * File            : s2u.c
 * Version         : $Id$
 * Authors         : Stew Benedict and Frederic Lepied
 * Created On      : Mon Aug  2 08:28:19 2004
 * Purpose         : read dbus messages addressed to the user and take
 *                 appropriate actions.
 *===========================================================================*/

/***************************************************************************
 *
 * Copyright (C) 2004,2005,2009 Mandrakesoft
 * Copyright (C) 2005 Mandriva
 *
 * Stew Benedict, <sbenedict@mandriva.com>
 * Frederic Lepied, <flepied@mandriva.com>
 * Eugeni Dodonov <eugeni@mandriva.com>
 *
 * code borrowed/adapted from the hal project -
 *                  http://www.freedesktop.org/Software/hal
 *
 * Licensed under the GNU General Public License version 2.0
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 **************************************************************************/

static const char rcs_id[] = "$Id$";
static const char compile_id[] = "$Compile: " __FILE__ " " __DATE__ " " __TIME__ " $";

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <getopt.h>
#include <pwd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <errno.h>
#include <signal.h>
#include <gdk/gdk.h>
#include <libnotify/notify.h>

#include <dbus/dbus.h>
#include <dbus/dbus-glib.h>
#include <dbus/dbus-glib-lowlevel.h>

static DBusConnection *dbus_connection;
static gchar *cookie = NULL;
static NotifyNotification *n;

/** Print out program usage.
 *
 */
static void
usage ()
{
    g_printerr ("\n" "usage : s2u [--daemon=yes|no] [--help] [--debug]\n");
    g_printerr (
         "\n"
         "        --daemon=yes|no    Become a daemon\n"
         "        --debug            Print debug informations\n"
         "        --help             Show this information and exit\n"
         "\n"
         "s2u monitors messages through D-BUS,\n"
         "taking appropraite actions.\n");
}


/** If #TRUE, we will daemonize */
static dbus_bool_t opt_become_daemon = FALSE;

static void
die (const char *message)
{
  g_printerr ("*** %s", message);
  exit (1);
}

/** Message handler for method invocations. All invocations on any object
 *  or interface is routed through this function.
 *
 *  @param  connection          D-BUS connection
 *  @param  message             Message
 *  @param  user_data           User data
 *  @return                     What to do with the message
 */

static DBusHandlerResult
filter_function (DBusConnection * connection,
         DBusMessage * message, void *user_data)
{

/*
    INFO (("obj_path=%s interface=%s method=%s",
    dbus_message_get_path(message),
    dbus_message_get_interface(message),
    dbus_message_get_member(message)));
*/

    /* message, interface, method */
    if (dbus_message_is_signal (message,
                     "com.mandriva.user",
                     "message") &&
        strcmp (dbus_message_get_path (message),
            "/com/mandriva/user") == 0) {
      gchar* args[] =
        {
          "/usr/bin/xauth",
          "add",
          NULL,
          ".",
          NULL,
          NULL
        };

      args[2] = getenv("DISPLAY");
      args[4] = cookie;

      g_print("s2u: detected hostname change; setting cookie to '%s' for %s.\n", cookie, args[2]);

      g_spawn_async("/", args, NULL, 0, NULL, NULL, NULL, NULL);

      return DBUS_HANDLER_RESULT_HANDLED;
    } else if (dbus_message_is_signal (message,
                     "com.mandriva.user",
                     "updatemenu")) {

          g_print("s2u: updatemenu signal received.\n");

      g_spawn_command_line_async("/etc/X11/xinit.d/menu", NULL);

      return DBUS_HANDLER_RESULT_HANDLED;
    } else if (dbus_message_is_signal (message,
                     "com.mandriva.user",
                     "security_notification")) {
        /* msec */
        char *string;
        DBusError error;
        dbus_error_init(&error);
        if (dbus_message_get_args (message,
                    &error,
                    DBUS_TYPE_STRING, &string,
                    DBUS_TYPE_INVALID)) {
            n = notify_notification_new("MSEC", string, GTK_STOCK_INFO, NULL);
            if (!notify_notification_show (n, NULL)) {
                g_printerr("notify_notification_show: failed to show notification\n");
            }
            g_object_unref(G_OBJECT(n));
        }
        else {
            fprintf (stderr, "an error occurred: %s\n", error.message);
        }
        dbus_error_free(&error);
        return DBUS_HANDLER_RESULT_HANDLED;
    } else {
        return DBUS_HANDLER_RESULT_NOT_YET_HANDLED;
    }
}


static void
service_dbus_init (void)
{
    GError *error = NULL;

    dbus_connection = dbus_g_connection_get_connection (dbus_g_bus_get (DBUS_BUS_SYSTEM, &error));
    if (dbus_connection == NULL) {
        g_printerr ("dbus_bus_get(): %s", error->message);
        exit (1);
    }

         dbus_bus_add_match (dbus_connection,
                             "type='signal',"
                             "interface='com.mandriva.user'",
                NULL);


    dbus_connection_add_filter (dbus_connection, filter_function, NULL,
                    NULL);

}

int
main (int argc, char *argv[])
{
    GMainLoop *loop;
    gchar* args[] = {
        "/usr/bin/xauth",
        "list",
        NULL,
        NULL
    };
    gint in;
    gchar result[255];
    gchar* idx;
    dbus_bool_t opt_debug = FALSE;

    while (1) {
        int c;
        int option_index = 0;
        const char *opt;
        static struct option long_options[] = {
            {"daemon", 1, NULL, 0},
            {"help", 0, NULL, 0},
            {"debug", 0, NULL, 0},
            {NULL, 0, NULL, 0}
        };

        c = getopt_long (argc, argv, "",
                 long_options, &option_index);
        if (c == -1)
            break;

        switch (c) {
        case 0:
            opt = long_options[option_index].name;

            if (strcmp (opt, "help") == 0) {
                usage ();
                return 0;
            } else if (strcmp (opt, "daemon") == 0) {
                if (strcmp ("yes", optarg) == 0) {
                    opt_become_daemon = TRUE;
                } else if (strcmp ("no", optarg) == 0) {
                    opt_become_daemon = FALSE;
                } else {
                    usage ();
                    return 1;
                }
            } else if (strcmp (opt, "debug") == 0) {
                opt_debug = TRUE;
            }
            break;

        default:
            usage ();
            return 1;
            break;
        }
    }

    memset(result, 0, sizeof(result));

    args[2] = getenv("DISPLAY");

    g_spawn_async_with_pipes("/", args, NULL, 0, NULL, NULL, NULL,
                 NULL, &in, NULL, NULL);
    if (read(in, result, sizeof(result)) <= 0) {
        die("unable to read X11 cookie");
    } else {
        close(in);

        idx = rindex(result, ' ');
        if (idx == NULL) {
            die ("unable to read X11 cookie");
        }
        cookie = g_strdup(idx+1);
        cookie[strlen(cookie) - 1] = '\0';
        if (opt_debug) {
            g_print("%s: cookie for %s = '%s'\n", argv[0], args[2], cookie);
        }
    }

    if (opt_become_daemon) {
        int child_pid;
        int dev_null_fd;

        if (chdir ("/") < 0) {
            g_printerr ("Could not chdir to /, errno=%d", errno);
            return 1;
        }

        child_pid = fork ();
        switch (child_pid) {
        case -1:
            g_printerr ("Cannot fork(), errno=%d", errno);
            break;

        case 0:
            /* child */

            dev_null_fd = open ("/dev/null", O_RDWR);
            /* ignore if we can't open /dev/null */
            if (dev_null_fd > 0) {
                /* attach /dev/null to stdout, stdin, stderr */
                dup2 (dev_null_fd, 0);
                if (opt_debug == FALSE) {
                        dup2 (dev_null_fd, 1);
                    dup2 (dev_null_fd, 2);
                }
            }

            umask (022);
            break;

        default:
            /* parent */
            exit (0);
            break;
        }

        /* Create session */
        setsid ();
    }

    gdk_init(&argc, &argv);

    /* init libnotify */
    notify_init("s2u");

    loop = g_main_loop_new (NULL, FALSE);

    /* set up the dbus services */
    service_dbus_init ();

    /* run the main loop and serve clients */

    g_main_loop_run (loop);

    return 0;
}

/*
 * Local variables:
 * mode: c
 * End:
 *
 * s2u.c ends here
 */
